packages/jsii/lib/compiler.ts

```diff
@@ -145,9 +159,102 @@ export class Compiler implements Emitter {
+            }
+        }
+        LOG.debug(`Creating or updating ${colors.blue(this.configPath)}`);
+        await fs.writeJson(this.configPath, this.typescriptConfig, {  replacer: utils.filterEmpty, spaces: 4 });
+    }
+
+    /**
+     * Find all dependencies that look like TypeScript projects.
+     *
+     * Enumerate all dependencies, if they have a tsconfig.json file with
+     * "composite: true" we consider them project references.
+     *
+     * Unfortunately it doesn't seem like the TypeScript compiler itself
+     * resolves transitive references in a way that
+     */
+    private async findProjectReferences(): Promise<string[]> {
+        const packageJsonPath = path.join(this.options.projectInfo.projectRoot, 'package.json');
+
+        // If there's no package file, don't do anything
+        if (!await fs.pathExists(packageJsonPath)) { return []; }
+        const pkg = require(packageJsonPath);
+
+        const ret = new Array<string>();
+        for (const dependencyMap of [pkg.dependencies, pkg.devDependencies]) {
+            if (dependencyMap === undefined) { continue; }
+
+            for (const depName of Object.keys(dependencyMap)) {
+                let tsconfigFile = path.join('node_modules', depName, 'tsconfig.json');
+                if (!await fs.pathExists(tsconfigFile)) { continue; }
+
+                // Resolve symlinks, otherwise the TypeScript compiler will not
+                // find the project the files belong to.
+                tsconfigFile = await fs.realpath(tsconfigFile);
+
+                const tsconfig = require(tsconfigFile);
+
+                // Add references to any TypeScript package we find that is 'composite' enabled.
+                // Make it relative.
+                if (tsconfig.compilerOptions && tsconfig.compilerOptions.composite) {
+                    ret.push(path.relative(this.options.projectInfo.projectRoot, path.dirname(tsconfigFile)));
+                } else {
+                    // Not a composite package--if this package is in a node_modules directory, that is most
+                    // likely correct, otherwise it is most likely an error (heuristic here, I don't know how to
+                    // properly check this).
+                    if (tsconfigFile.indexOf('node_modules') > -1) {
+                        LOG.warn('%s: not a composite TypeScript package, but it probably should be', path.dirname(tsconfigFile));
+                    }
+                }
+            }
+        }
+        return ret;
+    }
+
+    /**
+     * Find source files using the same mechanism that the TypeScript compiler itself uses.
+     *
+     * Respects includes/excludes/etc.
+     *
+     * This makes it so that running 'tsc' and running 'jsii' has the same behavior.
+     */
+    private async determineSources(files: string[]): Promise<void> {
+        this.rootFiles.splice(0);
```

Comment:
Would it make more sense to just return the list of files instead of mutate `rootFiles`? Seems unnecessarily convoluted.